DirectILLAutoProcess

darominski opened the pull request:
This PR implements new algorithm to automatize the data reduction for direct geometry TOF instruments at the ILL: PANTHER, SHARP, IN5, and retroactively also for IN4 and IN6. 

The algorithm connects already existing DirectILL algorithms: `DirectILLCollectData`, `DirectILLReduction`, `DirectILLIntegrateVanadium`, `DirectILLDiagnostics` in a way that frees user from connecting the inputs and outputs of the each individual step, as well as provide robust options for masking data and full attenuation treatment. The goal is to significantly shorten and simplify the scripts used by the non-expert user, who runs a standarized reduction workflow. For custom workflows, the individual calls to DirectILL family of algorithms in a script may still be needed.

**To test:**

<!-- Instructions for testing. -->
1. Code review.
2. Tests should be passing.
3. Run a minimal full powder reduction of PANTHER data using a script:

```
EmptyRuns = '9777'
VanaRuns = "9406"
SampleRuns = '9738, 9740, 9744'

EmptyName = 'MTCell19meV'
VanaName = "V19meV"
SampleName = 'He3C60'

Ei = 19
Elc = 104

geometry = {
            'Shape': 'FlatPlate',
            'Height': 4.0,
            'Width': 2.0,
            'Thick': 0.2,
            'Angle': 0.0,
            'Center': [0.0, 0.0, 0.0]
}

material = {
            'ChemicalFormula': 'C60-(He3)0.80',
            'ZParameter': 4.0,
            'UnitCellVolume': 2744.0,  # =14A**3 for one Cell CFC C60
            'SamplePackingFraction': 0.1
        }

DirectILLAutoProcess(
            Runs=EmptyRuns,
            OutputWorkspace=EmptyName,
            ProcessAs='Empty',
            ReductionType='Powder',
            IncidentEnergyCalibration="Energy Calibration ON",
            IncidentEnergy=Ei,
            ElasticChannelIndex=Elc,
            EPPCreationMethod='Calculate EPP',
            SaveOutput=False,
            ClearCache=True
)

# vanadium
DirectILLAutoProcess(
            Runs=VanaRuns,
            OutputWorkspace=VanaName,
            ProcessAs='Vanadium',
            ReductionType='Powder',
            EmptyContainerWorkspace=EmptyName,
            FlatBkg='Flat Bkg ON',
            IncidentEnergyCalibration="Energy Calibration ON",
            IncidentEnergy=Ei,
            ElasticChannelIndex=Elc,
            EPPCreationMethod='Calculate EPP',
            SaveOutput=False,
            ClearCache=True
)

# sample
DirectILLAutoProcess(
            Runs=SampleRuns,
            OutputWorkspace=SampleName,
            ProcessAs='Sample',
            ReductionType='Powder',
            VanadiumWorkspace=VanaName,
            FlatBkg='Flat Bkg OFF',
            EmptyContainerWorkspace=EmptyName,
            EPPCreationMethod='Calculate EPP',
            ElasticChannelIndex=Elc,
            MaskedAngles=[0, 20],
            IncidentEnergy=Ei,
            IncidentEnergyCalibration="Energy Calibration ON",
            AbsorptionCorrection='Fast',
            SelfAttenuationMethod='MonteCarlo',
            SampleMaterial=material,
            SampleGeometry=geometry,
            SaveOutput=False,
            ClearCache=True
)

```

You can change parameters for calibrations, and masking options from the provided parameters. There should be no issues.
4. Reduce the data from the example above in DrILL by filling out the table, using provided rundex.
5. Single crystal example:

```
ei = 19.03
elp = 104
Load(Filename='Vnorm012315.nxs', OutputWorkspace='Norm')
GroupWorkspaces(InputWorkspaces='Norm', OutputWorkspace='NormVana')
eBins = [-2., 0.19, 16]  # inelastic
group_by = 4

output_name = 'panther_236'
DirectILLAutoProcess(
            Runs='13806, 13807',
            OutputWorkspace=output_name,
            ProcessAs='Sample',
            ReductionType='SingleCrystal',
            VanadiumWorkspace='NormVana',
            MaskWorkspace='Mask212.xml',
            IncidentEnergyCalibration="Energy Calibration ON",
            IncidentEnergy=ei,
            ElasticChannelIndex=elp,
            EnergyRebinningParams=eBins,
            FlatBkg='Flat Bkg ON',
            GroupDetBy=group_by,
            ClearCache=True,
            SaveOutput=False
)
```
Fixes #28161 .

Release notes and documentation updated. 

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

darominski committed:
[DirectILLAuto] First list of properties

darominski committed:
[DirectILLAuto] Validation of basic properties

darominski committed:
Added cadmium absorber to inputs

darominski committed:
Set default value for masked tubes property

darominski committed:
Changed material setting properties to PropertyManagerProperty

darominski committed:
Added setUp, collect data, and mock call to process_sample

darominski committed:
Added support for IN4, IN5, IN6, PANTHER, and SHARP

darominski committed:
Added clear cache property

darominski committed:
Added CadmiumWorkspace to correction ws group

darominski committed:
Added processing vanadium

darominski committed:
Added method for intermediate ws removal

darominski committed:
Uniform naming scheme for reduced vanadium data

darominski committed:
Filled out single crystal sample processing

This includes extracting proper vanadium intergral and diagnostics workspaces, and improved setting of eBins class parameter in setUp.

darominski committed:
Corrected validation and setting of Ei and EPP

darominski committed:
Merge branch 'master' of https://github.com/mantidproject/mantid into 28161_directillauto

darominski committed:
Corrected defaultValue for SampleShape

darominski committed:
Corrected order of vanadium corrections

darominski committed:
Added most basic powder reduction

darominski committed:
Corrected normalisation when vanadium integral is not provided

darominski committed:
Removed obligatory vanadium input for sample reduction

darominski committed:
Removed redundant sample shape property

darominski committed:
Removed obligation to set container material & geometry

darominski committed:
Added self-attenuation correction

darominski committed:
Added saving output as NXSPE files at each iteration step.

darominski committed:
Correct extracting flatbckgscaling from properties

darominski committed:
Removed redundant  energy range property.

It is only used when there is an issue in the initial energy or it's not present in the data, which
is outside of the scope of this algorithm.

darominski committed:
Connected sample angle offset with setting Psi in saving NXSPE

darominski committed:
Reorganized and centralized calls to cleanup method

darominski committed:
Corrected obtaining masking parameters

LoadMask should be called instead of LoadNexusProcessed, masking angles need to be obtained with .value
of the property, and the mask preparation can only be called after the instrument identity has been
fixed.

darominski committed:
Fixed checking if parameters are defined

darominski committed:
Corrected ElasticChannelEnergy property name to ElasticChannelIndex

darominski committed:
Naming consistency for binning properties

darominski committed:
Changed the way vanadium input is provided

Now the user can point to a single file instead of a WorkspaceGroup extant in the ADS. An attempt
is made to load that file, and in case it is a single 2D workspace, it is groupped to remain consistent
with the expected input.

darominski committed:
Removed check for workspace existence in the ADS before loading.

This removes the vulnerability that the file exist but other expected inputs
produced by the DirectILLCollectData do not, as well as one related to properly
setting of the used instrument.

darominski committed:
Added saving .nxs for powder reduction, .nxspe for sx

darominski committed:
Removed default value as empty dictionary for sample geometry.

This fixes found issues in communicating dictionary parameters through DrILL interface.

darominski committed:
Cast PropertyManager to dict for JSON serialisation in Drill.

PropertyManager from kernel is not JSON serializable and thus cannot
be saved in rundex correctly. A solution is to simply cast PropertyManager
to a python dictionary

darominski committed:
Clean up can be a static method

darominski committed:
Current output name list variable simplification

darominski committed:
Added grouping pixels functionality.

Pixels will be grouped according to the provided input, either
using the provided grouping pattern or by the given number of pixels.

darominski committed:
Changed type for EmptyContainer and Cadmium inputs

Since these will be provided as output from different calls to the DirectILLAutoProcess
they will be necessarily workspace groups.

darominski committed:
Added case for processing empty container and cadmium

darominski committed:
Implemented subtracting empty container counts

darominski committed:
Changed default property values for improved readability in DrILL

Default values for PropertyManagerProperty were removed altogether,
while Property.EMPTY_DBL which was the biggest negative double
were replaced with 0.0..

darominski committed:
Exposed groupping behaviour to the user

darominski committed:
Connected GroupingAngleStep and binning options to DirectILLReduction

darominski committed:
Added instrument-dependent look-up for sample position offset

darominski committed:
Connected DirectILLAutoProcess with DrILL

darominski committed:
Changed positiveInt validation to allow for 0 in channelIndex property

darominski committed:
Define properly parameters for 'fast' MC self-attenuation method

darominski committed:
Set optional reduction options via keyword parameters.

This replaces setting them directly, which would create a conflict,
as their default options may not match with the default in the called
algorithm. Now, they will be set only if modified by the user.

darominski committed:
Changed algorithm's super to DataProcessorAlgorithm

Previous super of PythonAlgorithm does not support copying properties
from other algorithms, which would be beneficiary for this algorithm.

darominski committed:
Exposed FlatBkg and AveragingWindow parameters to the user.

Also, made property names related to flat backgroun more in line with
the DirectILLCollectData.

darominski committed:
Split MaskThreshold into Min and Max threshold

Now it is possible to define both a minimal and maximal threshold,
while previously only a minimal threshold was available.

darominski committed:
Added option for detector grouping before flatbkg calculation.

It is now possible to group detectors horizontally (at the same vertical
position and between different tubes) and vertically (in the same tube)
to increase the statistics for the estimate of the flat background.

darominski committed:
Corrected validation for min/max threshold

darominski committed:
Corrected powder workflow when vanadium is not provided

darominski committed:
Exposed horizontal/vertical grouping to the auto-process algorithm

This commit renames GroupPixelsBy and connects properties for
horizontal and vertical grouping in DirectILLCollectData.
GroupDetBy is used solely to decrease granularity of data and groups pixels vertically,
inside of a tube. It worth to keep it separate from horizontal/vertical grouping intended
for increasing statistics for the flat background calculation, as in the latter case
the granularity reduction may be much larger than desired for the final data.

darominski committed:
Added imposing a name scheme on output workspaces.

The new naming scheme adds the name of the output workspace as a suffix,
followed by the current name of the workspace at the end of the reduction,
followed by 'Ei' and the initial energy value, and in case of powder reduction
technique, also the 'T' and sample temperature.

darominski committed:
Added loading flat bkg source if it's not in ADS

darominski committed:
Added summary report at the end of reduction.

The report contains the first and last numor of the sample runs,
all input workspaces (cadmium, empty container, vanadium, mask, flat bkg),
and experiment variables: reduction type, processed as, incident energy,
and sample temperature.

darominski committed:
Removed obsolete comment about saving in DrILL

darominski committed:
Re-organised method order to alphabetical one

darominski committed:
Added progress reporting

darominski committed:
Corrected storing temperatures in case there is more than one.

For powder samples it can happen that the temperature of the input is an ndarray
instead of a single value. Then, this special treatment is necessary.

darominski committed:
SofTEnergy will be now stored as nxspe instead of nxs

darominski committed:
Corrected workflow with masking angle range

darominski committed:
Corrected setting masking criterions.

Previous implementation was not mathematically correct and also over-complicated.

darominski committed:
ElasticChannelWorkspace set only when defined by the user

in the call to DirectILLCollectData

darominski committed:
Removed hard-coding parameters in case of Vanadium data collection

darominski committed:
Added matching of self-attenuation workspace

In the powder reduction the bin edges are sometimes slightly different which does not agree
with the Divide algorithm. This change alleviates the found issue.

darominski committed:
Re-use already existing numor variable

darominski committed:
The chosen log temperature shall be now setpoint rather than measured

darominski committed:
Corrected setting output_samples list for Empty and Cadmium

darominski committed:
Removed redundant special case in vana for IN5

darominski committed:
Scale empty container only when scaling factor != 1

darominski committed:
Allow self-attenuation correction also for Vanadium and SX sample

darominski committed:
Vanadium diagnostic workspace is now saved via SaveMask

darominski committed:
Added cadmium subtraction

darominski committed:
Corrected workflow around masking

No masking workspace is created, and thus mask_ws shall be None when the only masking
is done via thresholds.

darominski committed:
Unit test for DirectILLAutoProcess

The test covers basic reductions of  empty container, vanadium, and single crystal
using PANTHER data. Powder and more complex modes will be placed as a system test
due to execution time requirements"

darominski committed:
[DirectILL] Re-organised elastic peak calibration parameters

Removed redunant ElasticPeakCalibration bool, exposed ElasticChannel
and EPPCreationMethod from DirectILLCollectData to the user, modified
parameters in DrILL and changed group contents of the AutoProcess algorithm
properties.

darominski committed:
[DirectILL] EP ws connected to DirectILLDiagnostics in vana reduction

darominski committed:
[DirectILL] Added first draft of AutoProcess documentation

darominski committed:
[DirectILL] Removed deprecated parameters from the unit test

darominski committed:
[DirectILL] Brought back set for beam stop diagnostics

darominski committed:
Merge remote-tracking branch 'origin/main' into 28161_directillauto

darominski committed:
[DirectILL] System test for PANTHER powder AutoProcess

The standard reduction includes empty container, vanadium, and sample measured at
three different temperatures.

darominski committed:
[DirectILL] Fixed max range in grouping by pattern

darominski committed:
[DirectILL] Changed no of spectra after changes in main branch

darominski committed:
[DirectILL] Changed custom ei calibration property to a copy

DirectILLAutoProcess will now use the same property as DirectILLCollectData
to avoid any issues with compatibility and default parameters.

darominski committed:
[DirectILL] Changed custom E binning property to copies

DirectILLAutoProcess now copies EnergyRebinning and EnergyRebinningParams
from DirectILLReduction in place of having only one manual binning option.

darominski committed:
[DirectILL] Allow mask ws to be a nexus file

Previously a mask workspace could be only an .xml file loaded with LoadMask
algorithm, while now it can also be a .nxs file loaded with LoadNexus.

darominski committed:
[DirectILL] Corrections to vanadium workflow in AutoProcess

Masks will be now applied also to vanadium and erroneous setting of
IntegratedVanadiumWorkspace is removed.

darominski committed:
[DirectILL] Improved naming for groupped flat bkg

in DirectILLCollectData. Previously, there was one underscore too many and also the name
of the flat background workspace created by first grouping detectors was not properly set
to the user selected name via property OutputFlatBkgWorkspace.

darominski committed:
[DirectILL] Improved progress report message

Grouping outputs is more informative than the previous message.

darominski committed:
[DirectILL] Removed saving DirectILLAutoProcess output as .nxspe

darominski committed:
[DirectILL] Corrected check of container geometry property

It was incorrectly checking for the sample geometry instead.

darominski committed:
[DirectILL] Corrected removing 'raw' from the MT name.

This removes the redundant _ from the set output name in empty container reduction.

darominski committed:
[DirectILL] Corrected check for existence of vanadium integral ws

darominski committed:
[DirectILL] Connect properly absolute unit normalisation

AbsoluteUnitNormalisation was not properly connected between DirectILLAutoProcess and DirectILLReduction,
this is now amended.

darominski committed:
[DirectILL] AutoProcess system test with single crystal PANTHER exp

darominski committed:
[DirectILL] Updated AutoProcess docs, added workflow diagrams

darominski committed:
[DirectILL] Updated technique docs with AutoProcess algorithm details

darominski committed:
[DirectILL] Release note for the AutoProcess algorithm

darominski committed:
[DirectILL] Release note for pixel counts averaging for flat bkg

gvardany reviewed (commented):
Just a few comments on the code and tests.
I run the provided example for powder.
It run through and produced workspaces as in the attachment.
![Screenshot 2022-02-07 at 13 18 06](https://user-images.githubusercontent.com/16865759/152786818-48775664-a980-49a8-b07e-7fbfd20efdf0.png)

There are a few issues with this.
First and most important, is that the SofQW workspaces are not correct. 
They have q as x-axis, but not deltaE as y-axis.
Are you sure those are the right outputs as produced by `DirectILLReduction`?
Similarly, the y-axis of SofTW is not 2 theta in degrees, but just spectrum number.

Then, for vanadium, the naming convention looks a bit strange.
The run numbers are in a different place as compared to the sample outputs.
Then, there is SofQ, which I don't understand. Is this really supposed to be the diffraction pattern, i.e. integrated over all deltaE bins?

There is a dangling workspace, called `corrected` which doesn't belong to any group. If clear cache is true by default, I wasn't expecting it.

Then, if self attenuation corrections are requested (no matter the method), I think we should output also the corrections (the workspace group containing up to 4 terms), even with the clear cache enabled. Those are long to compute, and useful to look at for diagnostics.

Also, I'd suggest adding the units to the observables in the workspace names, just to be more informative and also consistent with other techniques, such as SANS.
For example `.....Ei19mev_T20K` looks much better than now.

Finally, could you provide an example with the single crystal as well, containing at least 2 sample angles?

darominski committed:
[DirectILL] Improved consistency of vanadium ws naming

SofQ was renamed to a correct SofQW, and now numor is placed after the reduction
process, to be consistent with other reduction modes.

darominski committed:
[DirectILL] Added units to T, Ei in ws naming scheme

darominski commented on Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/DirectILLAutoProcessTest.py:
```diff
@@ -0,0 +1,107 @@
+        self._check_output(mtd[vanadium_name][1], 226, 669, True, "Energy transfer", "DeltaE", "Scattering angle",
+                           "Degrees")  # S(2theta, w)
+
+    def test_sample_single_crystal(self):
```
Absolutely, I can. I put the sample powder reduction only in the system test due to its length. The unit test takes currently about 15 s, with the powder test included it doubles.  If you think that's fine, I am going to include it.

gvardany replied:
I think this is critical enough to pay the price. 30 second won't make it the slowest test.

darominski replied:
It's added then.

darominski committed:
[DirectILL] Removed redundant option from DirectTOF Drill save options

darominski committed:
[DirectILL] SingleCrystal system test processes two samples now

darominski commented on Testing/SystemTests/tests/framework/DirectILLAutoProcessTest.py:
```diff
@@ -0,0 +1,170 @@
+
+        output_name = 'panther_236'
+        DirectILLAutoProcess(
+            Runs='13806',
```
Done. Also, example above also includes reduction of single crystal PANTHER data with two different sample positions.

darominski committed:
[DirectILL] Delete redundant 'corrected' ws produced by PP algorihtm

darominski committed:
[DirectILL] Suppressed deleting PaalmanPings correction workspaces

darominski committed:
[DirectILL] Added powder reduction to autoprocess unit tests

darominski committed:
[DirectILL] Specified meaning of multiple numors in docs

Co-authored-by: Gagik Vardanyan <vardanyan@ill.fr>

darominski committed:
Merge branch 'main' into 28161_directillauto

darominski committed:
[DirectILL] Strenghed the AutoProcess use-case in the docs

darominski committed:
[DirectILL] Fixed tests broken by merge

darominski committed:
[DirectILLAuto] Replaced symbolic link with repeated definition

darominski committed:
[DirectILL] Corrected empty subtraction and updated doc tests

Now, the empty container workspace will be matched with the appropriate sample index that is
being reduced. Also, in case of inconsistent X-axis binning, a RebinToWorkspace is now
called to amend this issue.

darominski committed:
[DirectILL] Re-organized workflow around grouping detectors.

Grouping now takes place at a different point for powder and single crystal
sample reduction, while vanadium is currently not groupped at all.

darominski committed:
[DirectILL] Disabled grouping for powder sample reduction.

There are issues in sample and vanadium reduction when the input
to DirectILLReduction is groupped, there is a confusion in naming
of components (as they have now composite names made of all groupped components)
in the DetectorEfficiencyCorUser, which leads to an exception in the reduction.

darominski committed:
[DirectILL] Removed repeated setting of GroupingAngleStep

darominski committed:
[DirectILL] Moved generating grouping pattern to commons

There were two similar instances creating the grouping pattern, one in DirectILLCollecData,
and the other in the DirectILLAutoProcess algorithm. Currently, there is only one instance
of the code and is placed in the DirectILL_common. A series of unit tests has been added
to ensure patterns are created properly.

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/DirectILLCollectData.py:
```diff
@@ -738,6 +786,58 @@ def _flatBkgMon(self, monWS):
         self._cleanup.cleanup(monWS)
         return bkgSubtractedMonWS
 
+    def _get_grouping_pattern(self, ws):
```
It had only slight differences, and could have been generalized into a method in `DirectILL_common`, which I have just done. Also, I added unit tests for both intra-tube and between-tubes grouping pattern creation for IN5, PANTHER, and SHARP.

darominski commented:
> Just a few comments on the code and tests. I run the provided example for powder. It run through and produced workspaces as in the attachment. ![Screenshot 2022-02-07 at 13 18 06](https://user-images.githubusercontent.com/16865759/152786818-48775664-a980-49a8-b07e-7fbfd20efdf0.png)
> 
> There are a few issues with this. First and most important, is that the SofQW workspaces are not correct. They have q as x-axis, but not deltaE as y-axis. Are you sure those are the right outputs as produced by `DirectILLReduction`? Similarly, the y-axis of SofTW is not 2 theta in degrees, but just spectrum number.
> 
> Then, for vanadium, the naming convention looks a bit strange. The run numbers are in a different place as compared to the sample outputs. Then, there is SofQ, which I don't understand. Is this really supposed to be the diffraction pattern, i.e. integrated over all deltaE bins?
> 
> There is a dangling workspace, called `corrected` which doesn't belong to any group. If clear cache is true by default, I wasn't expecting it.
> 
> Then, if self attenuation corrections are requested (no matter the method), I think we should output also the corrections (the workspace group containing up to 4 terms), even with the clear cache enabled. Those are long to compute, and useful to look at for diagnostics.
> 
> Also, I'd suggest adding the units to the observables in the workspace names, just to be more informative and also consistent with other techniques, such as SANS. For example `.....Ei19mev_T20K` looks much better than now.
> 
> Finally, could you provide an example with the single crystal as well, containing at least 2 sample angles?

Thank you for the review.

I corrected the naming scheme, and it is now both consistent for sample, empty, and vanadium, and naming also contains the relevant unit for parameters used (K, meV).

The issue with S (Q, w) was found in the grouping. Now, the grouping is performed only for the empty and single crystal sample, which is the workflow this feature was requested. There is no workflow where grouping would be of great use in the powder case, apart from increasing statistics of the flat background estimation, which is a separate thing. In any case, this issue has been fixed.

The dangling 'corrected' actually comes from `PaalmanPingsMonteCarloAbsorption`, but it is properly cleaned up now. Also, the absorption correction workspaces are kept, even if `ClearCache` is checked, as you suggested.

I updated the example of this PR to include the single crystal reduction, and expanded the system test to include two numors.

Other comments that were dispersed through the code have also been answered now. Could have a second look at this?

gvardany commented on docs/source/techniques/DirectILL.rst:
```diff
@@ -705,4 +743,101 @@ Output:
+
+.. testcode:: FullExampleAutoProcess
+
+    vanadium_runs = 'ILL/IN4/085801-085802'
```
Just a nitpick, if the `ILL/IN4` is added as a subdirectory, can we omit them in these paths to make it more appealing to the eye? Also `.nxs` extensions can be skipped.

gvardany commented:
Thanks for updates, it is much better.

> The issue with S (Q, w) was found in the grouping. Now, the grouping is performed only for the empty and single crystal sample, which is the workflow this feature was requested. There is no workflow where grouping would be of great use in the powder case, apart from increasing statistics of the flat background estimation, which is a separate thing. In any case, this issue has been fixed.

The axis of SQW is fixed now, however, I am not sure I understand the explanation. Detector grouping is irrelevant/plain wrong to run in Q space, as there is no longer per-pixel information available. In case you need to group data in reduced space, you should resort to Rebin, Rebin2D, etc. Can you confirm that in no scenario (powder, crystal, empty, vana, sample) there is no detector grouping attempted on reduced (i.e. Q converted) data?

I also noticed that in the provided example for powder, the attenuation correction workspace group is not present. However, it contains only one workspace. Is this because container geometry and material were not provided (even though container runs were)? Could you perhaps add container geometry and material in the respective test, and make sure that the attenuation correction workspace group contains all the 4 terms in PP formalism?

darominski committed:
[DirectILL] Removed redundant paths and extensions

from DirectILLAutoProcess reduction example in the technique documentation.

darominski commented:
> Thanks for updates, it is much better.
> 
> > The issue with S (Q, w) was found in the grouping. Now, the grouping is performed only for the empty and single crystal sample, which is the workflow this feature was requested. There is no workflow where grouping would be of great use in the powder case, apart from increasing statistics of the flat background estimation, which is a separate thing. In any case, this issue has been fixed.
> 
> The axis of SQW is fixed now, however, I am not sure I understand the explanation. Detector grouping is irrelevant/plain wrong to run in Q space, as there is no longer per-pixel information available. In case you need to group data in reduced space, you should resort to Rebin, Rebin2D, etc. Can you confirm that in no scenario (powder, crystal, empty, vana, sample) there is a detector grouping attempted on reduced (i.e. Q converted) data?
> 
> I also noticed that in the provided example for powder, the attenuation correction workspace group is now present. However, it contains only one workspace. Is this because container geometry and material were not provided (even though container runs were)? Could you perhaps add container geometry and material in the respective test, and make sure that the attenuation correction workspace group contains all the 4 terms in PP formalism?

Completely agreed that grouping after conversion to Q space makes no sense, and is not attempted. However, there would be an issue in `DetectorEfficiencyCorUser` if grouping had been done before `DirectILLReduction` was called, due to mentioned confusion of component names and fetching formula for the efficiency calculation. This runs deeper that this PR and should be thus fixed separately, I shall try to make an issue with a minimal example.

Yes, in the examples I had in hand there was no geometry provided for the container. Checking this out lead me to uncover issues in the `PaalmanPingsMonteCarloAbsorption`, where creating beam profile seem to always fail when sparse instrument is used, and effects of using `SetBeam` to set a beam geometry also seem to be always ignored by this algorithm. This definitely requires further investigation, but I would say it goes beyond this PR.

gvardany commented:
Ok, then please create the respective issues, to be handled quickly. For the former, it looks like you might need to move the grouping inside `DirectILLReduction`, in any case it must be before SofQW. I still don't what DetectorEfficiencyCorUser has to do with the grouping. In any case, I agree that this PR can advance as is.

darominski committed:
[DirectILL] Fixed naming scheme of rebinned PPMC params

A call to RebinToWorkspace was renaming the ass, acc, etc. workspaces
to a format not compatible with ApplyPaalmanPingsCorrection. This fix
preserves the desired formatting while still performing the needed
rebinning.
